src/ClientLease.cc

```diff
@@ -0,0 +1,108 @@
+/* Copyright (c) 2014-2015 Stanford University
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "ClientLease.h"
+#include "LeaseCommon.h"
+#include "RamCloud.h"
+
+namespace RAMCloud {
+
+/**
+ * Constructor for ClientLease.
+ */
+ClientLease::ClientLease(RamCloud* ramcloud)
+    : Dispatch::Timer(ramcloud->clientContext->dispatch)
+    , ramcloud(ramcloud)
+    , lease({0, 0, 0})
+    , lastRenewalTimeCycles(0)
+    , leaseTermElapseCycles(0)
+    , renewLeaseRpc()
+{}
+
+/**
+ * Return a valid client lease.  If a valid client lease has not been issued or
+ * the lease is about to expire, this method will block until it is able to
+ * return a valid client lease.
+ */
+WireFormat::ClientLease
+ClientLease::getLease()
+{
+    // If the lease renewal timer is not running, schedule it.  The lease
+    // is normally renewed in the background in parallel with other operations
+    // if possible.
+    if (!isRunning()) {
+        start(0);
+    }
+
+    // Block waiting for the lease to become valid.
+    while (Cycles::rdtsc() > leaseTermElapseCycles) {
+        ramcloud->poll();
+    }
+
+    return lease;
+}
+
+/**
+ * This method incrementally ensures that a the lease is valid; i.e. repeatedly
+ * calling this method will ensure the lease is valid.  This method is called
```

Comment:
I'm not sure exactly what the above sentence means. How "repeatedly" do I need to call this method in order to know that my lease is really valid? And, doesn't this method get invoked automatically, so clients don't need to worry about invoking it?